aptos-move/framework/aptos-stdlib/sources/cryptography/groth16.move

```diff
@@ -0,0 +1,169 @@
+    use aptos_std::groups::{BLS12_381_G1, BLS12_381_G2, deserialize_scalar, deserialize_element_uncompressed, BLS12_381_Gt};
+    use aptos_std::groups;
+
+    // Error codes
+    const E_NATIVE_FUN_NOT_AVAILABLE: u64 = 1;
+
+    /// A Groth16 verifying key.
+    struct VerifyingKey<phantom G1, phantom G2, phantom Gt> has drop {
+        alpha_g1: groups::Element<G1>,
+        beta_g2: groups::Element<G2>,
+        gamma_g2: groups::Element<G2>,
+        delta_g2: groups::Element<G2>,
+        gamma_abc_g1: vector<groups::Element<G1>>,
+    }
+
+    /// A Groth16 verifying key pre-processed for faster verification.
+    struct PreparedVerifyingKey<phantom G1, phantom G2, phantom Gt> has drop {
+        alpha_g1_beta_g2: groups::Element<Gt>,
+        gamma_g2_neg: groups::Element<G2>,
+        delta_g2_neg: groups::Element<G2>,
+        gamma_abc_g1: vector<groups::Element<G1>>,
+    }
+
+    /// A Groth16 proof.
+    struct Proof<phantom G1, phantom G2, phantom Gt> has drop {
+        a: groups::Element<G1>,
+        b: groups::Element<G2>,
+        c: groups::Element<G1>,
+    }
+
+    /// Create a new Groth16 verifying key.
+    public fun new_vk<G1,G2,Gt>(alpha_g1: groups::Element<G1>, beta_g2: groups::Element<G2>, gamma_g2: groups::Element<G2>, delta_g2: groups::Element<G2>, gamma_abc_g1: vector<groups::Element<G1>>): VerifyingKey<G1,G2,Gt> {
+        abort_if_feature_disabled();
+        VerifyingKey {
+            alpha_g1,
+            beta_g2,
+            gamma_g2,
+            delta_g2,
+            gamma_abc_g1,
+        }
+    }
+
+    /// Create a new pre-processed Groth16 verifying key.
+    public fun new_pvk<G1,G2,Gt>(alpha_g1_beta_g2: groups::Element<Gt>, gamma_g2_neg: groups::Element<G2>, delta_g2_neg: groups::Element<G2>, gamma_abc_g1: vector<groups::Element<G1>>): PreparedVerifyingKey<G1,G2,Gt> {
+        abort_if_feature_disabled();
+        PreparedVerifyingKey {
+            alpha_g1_beta_g2,
+            gamma_g2_neg,
+            delta_g2_neg,
+            gamma_abc_g1,
+        }
+    }
+
+    /// Pre-process a Groth16 verification key `vk` for faster verification.
+    public fun prepare_verifying_key<G1,G2,Gt>(vk: &VerifyingKey<G1,G2,Gt>): PreparedVerifyingKey<G1,G2,Gt> {
+        abort_if_feature_disabled();
+        PreparedVerifyingKey {
+            alpha_g1_beta_g2: groups::pairing<G1,G2,Gt>(&vk.alpha_g1, &vk.beta_g2),
+            gamma_g2_neg: groups::element_neg(&vk.gamma_g2),
+            delta_g2_neg: groups::element_neg(&vk.delta_g2),
+            gamma_abc_g1: vk.gamma_abc_g1,
+        }
+    }
+
+    /// Create a Groth16 proof.
+    public fun new_proof<G1,G2,Gt>(a: groups::Element<G1>, b: groups::Element<G2>, c: groups::Element<G1>): Proof<G1,G2,Gt> {
+        abort_if_feature_disabled();
+        Proof { a, b, c }
+    }
+
+    /// Verify a Groth16 proof.
+    public fun verify_proof<G1,G2,Gt>(vk: &VerifyingKey<G1,G2,Gt>, public_inputs: &vector<groups::Scalar<G1>>, proof: &Proof<G1,G2,Gt>): bool {
```

Comment:
So a pairing-product should be used here to speed up things, no?

Reply:
Fixed.